[Routing] Proactively checks constraints and complex parameters when building the DFA tree to trim unnecessary nodes

javiercn opened the pull request:
Due to the way we construct our DFA tree, several route patterns where there is a variable at the beginning of the route pattern in addition to a large set of different literals as the first segment of the route tree, results in massively large route tables (GB and more) many of them filled with nodes that we don't need.

The reason we don't need those nodes is because the literal they are hanging from won't match the constraint of the parameter or the shape of the complex parameter (or a constraint in a complex parameter part), hence even if we get to a candidate for that pattern, it will be discarded when we check constraints and complex parameters.

Instead of adding those nodes when building the DFA tree and evaluating the constraints and complex parameters on each request we evaluate the constraints and the complex parameters as we are building the DFA by evaluating the parent literal candidate against the constraint/complex parameter.

This results in DFAs with a much smaller number of nodes (several orders of magnitude depending on the case) and improves memory as well as startup time on these cases.

javiercn committed:
RouteTable size improvements

javiercn committed:
Add trimming for complex parameters

javiercn committed:
Update constraints to support trimming literal nodes

javiercn committed:
Add DFA matcher tests

javiercn committed:
Add DfaMatcherBuilder tests

javiercn committed:
Add complex parameter test

javiercn committed:
Add regression tests for well-known route patterns

javiercn commented on src/Http/Routing.Abstractions/src/ILiteralConstraint.cs:
```diff
@@ -0,0 +1,22 @@
+    /// When a parameter implements this interface, the router is able to optimize away some paths from the route table that don't match this constraint.
+    /// </remarks>
+    /// </summary>
+    public interface ILiteralConstraint : IParameterPolicy
+    {
+        /// <summary>
+        /// Determines whether the given <paramref name="literal"/> can match the constraint.
+        /// </summary>
+        /// <param name="parameterName">The parameter name we are currently evaluating.</param>
+        /// <param name="literal">The literal to test the constraint against.</param>
+        /// <returns><c>true</c> if the literal contains a valid value; otherwise, <c>false</c>.</returns>
+        bool MatchLiteral(string parameterName, string literal);
+    }
```
Need to come up with a better name for this, in essence we want to implement this on our constraints and let people implement it too. Other names are `IParameterNodeLiteralPolicy` or `IParameterNodeBuilderPolicy`.

The idea here is that we pass literal values when we are building the DFA tree and you get to decide whether they match or not. 

We also pass you the parameter name so that you can implement policies based on the parameter name even if you don't know the values ahead of time. For example, if the parameter name is controller or page, you can choose to never add any literal from the tree to mean something like, this parameter will never match any value from the list of literals in the tree.

javiercn replied:
I'm going to move this from the abstractions assembly to the implementation assembly, since this is mostly a DFA concern and rename it to IRouteParameterNodeFilterPolicy

Tratcher commented:
Issue #?

Who do you want to review this? I don't think Brennan or I have the context.

javiercn commented:
> Issue #?
> 
> Who do you want to review this? I don't think Brennan or I have the context.

@Tratcher  I've removed you from it. I'll ask James.

Here is the main issue:
https://github.com/dotnet/aspnetcore/issues/32781
Here are the related ones:
https://github.com/dotnet/aspnetcore/issues/23850
https://github.com/OData/AspNetCoreOData/issues/160
https://github.com/dotnet/aspnetcore/issues/30092

javiercn commented on src/Http/Routing.Abstractions/src/Properties/AssemblyInfo.cs:
```diff
@@ -10,7 +10,9 @@
 [assembly: TypeForwardedTo(typeof(EndpointMetadataCollection))]
+#pragma warning disable RS0016
 [assembly: TypeForwardedTo(typeof(RouteValueDictionary))]
+#pragma warning restore RS0016
```
VS and the build were complaining about this

javiercn committed:
Fix issue where the same candidate was being added twice

JamesNK commented on src/Http/Routing/src/Matching/DfaMatcherBuilder.cs:
```diff
@@ -264,10 +265,52 @@ public DfaNode BuildDfaTree(bool includeLabel = false)
+                                    var hasFailingPolicy = new bool[parent.Literals.Keys.Count];
+
+                                    // Whether or not all parameters have failed to meet at least one constraint.
+                                    for (var k = 0; k < parameterPolicyReferences.Count; k++)
```
nit: There are a lot of indexer variables: k, m, l, etc. I think they should have descriptive names.

k = parameterPolicyReferenceIndex, etc

Alternatively, split logic out into its own method.

javiercn replied:
Yeah, I'm going to break this method down, just didn't want to put that in the PR at first to avoid the confusion its going to cause.

javiercn replied:
I've cleaned up the implementation since the method was gigantic and had a lot of context associated with it. 

I've created a worker to hold the state and moved the construction to a method that is broken down by stage and segment type.

javiercn replied:
Added a bunch of comments along the way about how this works.

JamesNK commented on src/Http/Routing/src/Matching/DfaMatcherBuilder.cs:
```diff
@@ -264,10 +265,52 @@ public DfaNode BuildDfaTree(bool includeLabel = false)
+                                    for (var k = 0; k < parameterPolicyReferences.Count; k++)
+                                    {
+                                        var reference = parameterPolicyReferences[k];
+                                        var parameterPolicy = _parameterPolicyFactory.Create(parameterPart, reference);
```
When do constraints normally get created and where are they cached?

javiercn replied:
It's cached in the reference https://github.com/dotnet/aspnetcore/blob/main/src/Http/Routing/src/ParameterPolicyFactory.cs#L37

JamesNK commented on src/Http/Routing/src/Matching/DfaMatcherBuilder.cs:
```diff
@@ -288,7 +331,57 @@ public DfaNode BuildDfaTree(bool includeLabel = false)
+                                // that it will never be a match.
+                                foreach (var literal in parent.Literals.Keys)
+                                {
+                                    var routeValues = new RouteValueDictionary();
```
How much do we care about allocating this RVD every time vs reusing it?

davidfowl replied:
Reuse it if possible.

javiercn replied:
done

JamesNK commented on src/Http/Routing/src/Matching/DfaMatcherBuilder.cs:
```diff
@@ -264,10 +265,52 @@ public DfaNode BuildDfaTree(bool includeLabel = false)
+                                if (endpoint.RoutePattern.ParameterPolicies.TryGetValue(parameterPart.Name, out var parameterPolicyReferences))
+                                {
+                                    // The list of parameters that fail to meet at least one ILiteralConstraint.
+                                    var hasFailingPolicy = new bool[parent.Literals.Keys.Count];
```
stackalloc if `parent.Literals.Keys.Count` is smaller than 256? (which should be always, but allocate larger array just in case)

JamesNK replied:
Although, probably should only do that if this logic is in its own method. stackalloc in a loop doesn't behave well.

javiercn replied:
this is now in its own method

JamesNK reviewed (commented):
A review focusing on code.

I'll wait until we've talked through this before thinking about the overall solution.

javiercn commented on src/Http/Routing/src/Constraints/BoolRouteConstraint.cs:
```diff
@@ -43,5 +43,10 @@ public class BoolRouteConstraint : IRouteConstraint
+
+        bool ILiteralConstraint.MatchLiteral(string parameterName, string literal)
+        {
+            return bool.TryParse(literal, out _);
```
That's fair, I started calling the private method implementation from the constraint, but it was cumbersome. I'll put the code in a private method and call it from both places.

javiercn replied:
I've moved the code to a common method where it made sense

javiercn commented on src/Http/Routing/src/Matching/DfaMatcherBuilder.cs:
```diff
@@ -809,4 +902,44 @@ private static bool TryGetRequiredValue(RoutePattern routePattern, RoutePatternP
     }
+
+    [DebuggerDisplay("{" + nameof(GetDebuggerDisplay) + "(),nq}")]
+    internal struct WorkItem
```
I think so, I haven't dabbled with them so far, the tuple was making it hard for me to go through the algorithm, so I just did a VS refactor to turn it into a struct and that's what it came out.

javiercn replied:
![image](https://user-images.githubusercontent.com/6995051/122911073-c3528c00-d30b-11eb-84df-cddd550c65a7.png)

javiercn replied:
I'll make this a struct record when its feasible.

javiercn commented on src/Http/Routing/src/Matching/DfaMatcherBuilder.cs:
```diff
@@ -264,10 +265,52 @@ public DfaNode BuildDfaTree(bool includeLabel = false)
                             if (parent.Literals != null)
                             {
-                                nextParents.AddRange(parent.Literals.Values);
+                                // If the parameter contains constraints, we can be smarter about it and evaluate them while we build the tree.
+                                // If the literal doesn't match any of the constraints, we can prune the branch.
+                                // For example, for a parameter in a route {lang:length(2)} and a parent literal "ABC", we can check that "ABC"
+                                // doesn't meet the parameter constraint (length(2)) when building the tree, and avoid the extra nodes.
```
Fair enough, I didn't want to add too much stuff since this method is gigantic.

I'm going to break it into multiple methods to maintainability

javiercn committed:
tmp

javiercn committed:
Remove compat switch from 6.0

javiercn committed:
Cleanup, feedback, refactor the core algorithm

javiercn committed:
Remove conditional statements

javiercn committed:
Additional tests

javiercn committed:
Put checks in common private method

javiercn commented on src/Http/Routing/src/Matching/DfaMatcherBuilder.cs:
```diff
@@ -264,10 +265,52 @@ public DfaNode BuildDfaTree(bool includeLabel = false)
+                                    var l = 0;
+                                    foreach (var literal in parent.Literals.Values)
+                                    {
+                                        if (!hasFailingPolicy[l])
```
It was a bug, the test I was using was changing the first parameter :smile: I've added a couple of additional tests. Hopefully now that things are broken down its much clearer.

JamesNK commented on src/Http/Routing/src/Matching/DfaMatcherBuilder.cs:
```diff
@@ -186,131 +189,257 @@ public DfaNode BuildDfaTree(bool includeLabel = false)
+            {
+                // The list of parameters that fail to meet at least one ILiteralConstraint.
+                var hasFailingPolicy = parent.Literals.Keys.Count < 256 ?
+                    stackalloc bool [parent.Literals.Keys.Count] :
```
Use a constant size plus slice with stackalloc. It is faster to use a constant size because the compiler knows the exact stack size required and can make optimizations.

javiercn replied:
I've added a constant size and reduced this to 32 instead of 256, since nobody will have more than 32 parameters in 1 segment

JamesNK commented on src/Http/Routing/src/Matching/DfaMatcherBuilder.cs:
```diff
@@ -809,4 +938,45 @@ private static bool TryGetRequiredValue(RoutePattern routePattern, RoutePatternP
+    [DebuggerDisplay("{" + nameof(GetDebuggerDisplay) + "(),nq}")]
+    internal struct DfaBuilderWorkerWorkItem
+    {
+        public RouteEndpoint endpoint;
+        public int precedenceDigit;
+        public List<DfaNode> parents;
```
readonly + underscore prefix

JamesNK commented on src/Http/Routing/src/Matching/DfaMatcherBuilder.cs:
```diff
@@ -809,4 +938,45 @@ private static bool TryGetRequiredValue(RoutePattern routePattern, RoutePatternP
+            this.parents = parents;
+        }
+
+    public override bool Equals(object obj)
```
nit formatting

JamesNK commented on src/Http/Routing/src/Matching/DfaMatcherBuilder.cs:
```diff
@@ -186,131 +189,257 @@ public DfaNode BuildDfaTree(bool includeLabel = false)
+                var routeValues = new RouteValueDictionary();
+                foreach (var literal in parent.Literals.Keys)
+                {
+                    if (RoutePatternMatcher.MatchComplexSegment(segment, literal, routeValues))
```
Clear routeValues before each MatchComplexSegment call?

JamesNK replied:
never mind, I see it is cleared at the bottom

JamesNK commented on src/Http/Routing/src/Matching/DfaMatcherBuilder.cs:
```diff
@@ -186,131 +189,257 @@ public DfaNode BuildDfaTree(bool includeLabel = false)
+                    }
+                }
+
+                // Prepare the process the next stage.
```
```suggestion
                // Prepare to process the next stage.
```
?

javiercn replied:
done

JamesNK commented on src/Http/Routing/src/Matching/DfaMatcherBuilder.cs:
```diff
@@ -117,39 +93,66 @@ public DfaNode BuildDfaTree(bool includeLabel = false)
+            return root;
+        }
+
+        private struct DfaBuilderWorker
```
IMO this should be a class.
- It's quite a large struct
- It's mutable
- Only one is created

javiercn replied:
fair enough

javiercn committed:
Add additional regression test

javiercn committed:
PR feedback

javiercn committed:
Add additional comments about why we filter nodes

dougbu reviewed (dismissed):
New files need MIT license headers

davidfowl commented:
Is this change going in @javiercn ?

javiercn commented:
@davidfowl RC1

javiercn commented:
@davidfowl @JamesNK any further concern here?

I want to go through public API review with this today.

javiercn committed:
Updates from API review

javiercn committed:
Update license header

JamesNK commented on src/Http/Routing/src/Matching/DfaMatcherBuilder.cs:
```diff
@@ -808,5 +942,46 @@ private static bool TryGetRequiredValue(RoutePattern routePattern, RoutePatternP
             return !RouteValueEqualityComparer.Default.Equals(value, string.Empty);
         }
+
+        // TODO: Convert to record struct when available
+        [DebuggerDisplay("{" + nameof(GetDebuggerDisplay) + "(),nq}")]
+        private struct DfaBuilderWorkerWorkItem
```
Record struct now available?

javiercn replied:
I'll check for this and try out removing the #pragma

javiercn committed:
Fix breaking changes file, address remaining feedback

javiercn committed:
Apply workaround for
src\Http\Routing.Abstractions\src\Properties\AssemblyInfo.cs(13,12):
error RS0016: Symbol 'TryAdd' is not part of the declared API.
[]...Microsoft.AspNetCore.Routing.Abstractions.csproj

javiercn committed:
Fix baseline

javiercn committed:
Fix broken API baseline

javiercn committed:
Fix this for realsies

javiercn committed:
Another try
